cpp/src/scalar/scalar.cpp

```diff
@@ -446,4 +449,50 @@ list_scalar::list_scalar(cudf::column&& data,
 
 column_view list_scalar::view() const { return _data.view(); }
 
+struct_scalar::struct_scalar() : scalar(data_type(type_id::STRUCT)) {}
+
+struct_scalar::struct_scalar(table_view const& data,
+                             bool is_valid,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+  : scalar(data_type(type_id::STRUCT), is_valid, stream, mr), _data(data, stream, mr)
+{
+  CUDF_EXPECTS(
+    std::all_of(data.begin(), data.end(), [](column_view const& col) { return col.size() == 1; }),
+    "Struct scalar inputs must have exactly 1 row");
+
+  // validity pushdown
+  if (!is_valid) { superimpose_nulls(stream, mr); }
+}
+
+struct_scalar::struct_scalar(host_span<const column_view> data,
+                             bool is_valid,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+  : scalar(data_type(type_id::STRUCT), is_valid, stream, mr),
+    _data(table_view{std::vector<column_view>{data.begin(), data.end()}}, stream, mr)
+{
+  CUDF_EXPECTS(
+    std::all_of(data.begin(), data.end(), [](column_view const& col) { return col.size() == 1; }),
+    "Struct scalar inputs must have exactly 1 row");
+
+  // validity pushdown
+  if (!is_valid) { superimpose_nulls(stream, mr); }
```

Comment:
constructors have the same bodies, maybe extract into a private `init` function?